File: sdk/python/agentguard/policy.py
# agentguard/policy.py
# Policy YAML loader and validator for Python.
# Useful for:
#   - Validating a policy file before deploying it
#   - Generating policy files programmatically
#   - Testing what decision would be made for a given agent+tool

from __future__ import annotations
import os
from typing import Optional


def load_policy(path: str) -> dict:
    """
    Load and parse a policy YAML file.

    Args:
        path: Path to the policy.yaml file.

    Returns:
        dict: Parsed policy data.

    Raises:
        FileNotFoundError: If the file does not exist.
        ValueError: If the YAML is invalid.
    """
    try:
        import yaml
    except ImportError:
        raise ImportError(
            "PyYAML is required: pip install pyyaml"
        )

    if not os.path.exists(path):
        raise FileNotFoundError(f"Policy file not found: {path}")

    with open(path, "r") as f:
        data = yaml.safe_load(f)

    return data or {}


def validate_policy(policy: dict) -> list[str]:
    """
    Validate a policy dict against the AgentGuard schema.

    Returns:
        list[str]: List of validation errors.
                   Empty list = valid policy.

    Example:
        policy = load_policy("./policy.yaml")
        errors = validate_policy(policy)
        if errors:
            for e in errors:
                print(f"  ✗ {e}")
        else:
            print("  ✓ Policy is valid")
    """
    errors = []

    if "version" not in policy:
        errors.append("Missing required field: version")

    default = policy.get("default", "")
    if default not in ("allow", "block", "escalate", ""):
        errors.append(
            f"Invalid default decision: '{default}'. "
            "Must be 'allow', 'block', or 'escalate'."
        )

    agents = policy.get("agents", [])
    if not isinstance(agents, list):
        errors.append("'agents' must be a list")
        return errors

    seen_ids = set()
    for i, agent in enumerate(agents):
        if not isinstance(agent, dict):
            errors.append(f"agents[{i}]: must be a dict")
            continue

        aid = agent.get("id")
        if not aid:
            errors.append(f"agents[{i}]: missing 'id' field")
        elif aid in seen_ids:
            errors.append(f"agents[{i}]: duplicate id '{aid}'")
        else:
            seen_ids.add(aid)

        # Validate rule lists
        for rule_key in ("allow", "block", "escalate"):
            rules = agent.get(rule_key, [])
            if not isinstance(rules, list):
                errors.append(
                    f"agents[{i}].{rule_key}: must be a list"
                )
                continue
            for j, rule in enumerate(rules):
                if not isinstance(rule, dict) or "tool" not in rule:
                    errors.append(
                        f"agents[{i}].{rule_key}[{j}]: "
                        "must have a 'tool' field"
                    )

        spend = agent.get("spend_limit_daily_usd")
        if spend is not None and (
            not isinstance(spend, (int, float)) or spend < 0
        ):
            errors.append(
                f"agents[{i}].spend_limit_daily_usd: "
                "must be a non-negative number"
            )

    return errors


def simulate_decision(
    policy:   dict,
    agent_id: str,
    tool:     str,
) -> str:
    """
    Simulate what decision AgentGuard would make.
    Useful for testing policies without running the Go service.

    Returns: "allow", "block", or "escalate"
    """
    # Hardcoded irreversible tools always escalate
    _IRREVERSIBLE = {
        "delete", "delete_file", "delete_record", "drop_table",
        "drop_database", "truncate", "send_payment", "transfer_funds",
        "charge_card", "send_email", "send_sms", "execute_shell",
        "run_command", "exec", "bash", "eval", "deploy",
    }
    if tool.lower() in _IRREVERSIBLE:
        return "escalate"

    agents   = policy.get("agents", [])
    default  = policy.get("default", "block")
    tool_low = tool.lower()

    # Find matching agent policy
    matched  = None
    wildcard = None
    for a in agents:
        if a.get("id") == agent_id:
            matched = a
            break
        if a.get("id") == "*":
            wildcard = a

    active = matched or wildcard
    if not active:
        return default

    def matches(pattern: str) -> bool:
        return pattern == "*" or pattern.lower() == tool_low

    # Block takes priority
    for rule in active.get("block", []):
        if matches(rule.get("tool", "")):
            return "block"

    # Then escalate
    for rule in active.get("escalate", []):
        if matches(rule.get("tool", "")):
            return "escalate"

    # Then allow
    for rule in active.get("allow", []):
        if matches(rule.get("tool", "")):
            return "allow"

    return default

File: sdk/node/src/policy.ts
// src/policy.ts
// Policy loader and simulator for Node.js/TypeScript.

import { readFileSync } from 'fs'

export interface ToolRule {
  tool: string
}

export interface AgentPolicy {
  id:                        string
  allow?:                    ToolRule[]
  block?:                    ToolRule[]
  escalate?:                 ToolRule[]
  spend_limit_daily_usd?:    number
  irreversible_require_human?: boolean
}

export interface PolicyFile {
  version: string
  default: 'allow' | 'block' | 'escalate'
  agents:  AgentPolicy[]
}

export type Decision = 'allow' | 'block' | 'escalate'

const IRREVERSIBLE_TOOLS = new Set([
  'delete', 'delete_file', 'delete_record', 'drop_table',
  'drop_database', 'truncate', 'send_payment', 'transfer_funds',
  'charge_card', 'send_email', 'send_sms', 'execute_shell',
  'run_command', 'exec', 'bash', 'eval', 'deploy',
])

/**
 * Load a policy YAML file.
 * Requires 'js-yaml': npm install js-yaml
 */
export function loadPolicy(path: string): PolicyFile {
  let yaml: typeof import('js-yaml')
  try {
    yaml = require('js-yaml')
  } catch {
    throw new Error('js-yaml is required: npm install js-yaml')
  }
  const content = readFileSync(path, 'utf-8')
  return yaml.load(content) as PolicyFile
}

/**
 * Validate a policy object. Returns array of error strings.
 * Empty array = valid.
 */
export function validatePolicy(policy: PolicyFile): string[] {
  const errors: string[] = []

  if (!policy.version) errors.push('Missing required field: version')

  if (policy.default && !['allow','block','escalate'].includes(policy.default)) {
    errors.push(`Invalid default: '${policy.default}'`)
  }

  if (!Array.isArray(policy.agents)) {
    errors.push("'agents' must be an array")
    return errors
  }

  const seenIds = new Set<string>()
  policy.agents.forEach((agent, i) => {
    if (!agent.id) {
      errors.push(`agents[${i}]: missing 'id'`)
    } else if (seenIds.has(agent.id)) {
      errors.push(`agents[${i}]: duplicate id '${agent.id}'`)
    } else {
      seenIds.add(agent.id)
    }

    for (const key of ['allow', 'block', 'escalate'] as const) {
      const rules = agent[key]
      if (rules !== undefined && !Array.isArray(rules)) {
        errors.push(`agents[${i}].${key}: must be an array`)
      }
    }
  })

  return errors
}

/**
 * Simulate what decision AgentGuard would make.
 * Useful for unit-testing your policies.
 *
 * @example
 * const policy = loadPolicy('./policy.yaml')
 * expect(simulateDecision(policy, 'billing-agent', 'read_invoice'))
 *   .toBe('allow')
 */
export function simulateDecision(
  policy:  PolicyFile,
  agentId: string,
  tool:    string,
): Decision {
  const toolLow = tool.toLowerCase()

  // Hardcoded irreversible = always escalate
  if (IRREVERSIBLE_TOOLS.has(toolLow)) return 'escalate'

  const matched  = policy.agents.find(a => a.id === agentId)
  const wildcard = policy.agents.find(a => a.id === '*')
  const active   = matched ?? wildcard

  if (!active) return policy.default ?? 'block'

  const matches = (pattern: string) =>
    pattern === '*' || pattern.toLowerCase() === toolLow

  for (const r of active.block   ?? []) if (matches(r.tool)) return 'block'
  for (const r of active.escalate ?? []) if (matches(r.tool)) return 'escalate'
  for (const r of active.allow   ?? []) if (matches(r.tool)) return 'allow'

  return policy.default ?? 'block'
}
